### PR TITLE
[ADD][12.0] pos_meal_voucher

### PR DESCRIPTION
**Pull request by legalsylvain**
This module extend the Point of Sale Odoo module, regarding Meal Vouchers.

Meal voucher is a payment method, available in some countries (France, Belgium, Romania, ...) that allows customer to buy food products in grocery stores or pay in restaurants.

## Configuration

go to your products, and click on the 'Meal Voucher' checkbox, if your product can be paid with meal vouchers.

product_product_form

You can configure your product categories to have a default value for the products that belong to this category.

product_category_form

configures your Account journals, mentioning the type of Meal Voucher:

-Paper : the journal will be used when scanning meal voucher barcodes
- Dematerialized: the journal will be used for dematerialized meal vouchers
- Mixed: Specific configuration if your accountant want to use a single journal for Credit card AND dematerialized meal vouchers. In that case, the button of this journal will be duplicated, and an extra text can be set to display an alternative label.

account_journal_form

go your point of sale configuration form, and set the maximum amount allowed by ticket. (optional)

pos_config_form

## Usage

Open your Point of Sale

Cashier can see the food products, eligible for meal voucher payment, and see the total for Meal Voucher amount

front_ui_pos_order_screen

go to the payment screen

A Meal Voucher Summary is available:

front_ui_pos_payment_screen

If the amount received is too important, a warning icon is displayed

front_ui_pos_payment_screen_summary

If the cashier try to validate the order, a warning is also display, asking confirmation

front_ui_pos_payment_screen_warning

It is a non blocking warning, because we don't want to prevent an order to be done, if products are not correctly set, or if a recent law changed the maximum amount that can be used each day. (A recent case occured in France, during the Covid-19 pandemy)

### Note

A new barcode rule is available for Paper Meal Voucher of 24 chars:

...........{NNNDD}........

If you scan the following barcode 052566641320080017000000, a new payment line will be added, with an amount of 8,00€ (00800)

Full description : https://github.com/legalsylvain/pos/tree/12.0-ADD-pos_meal_voucher/pos_meal_voucher

CC : @sandiefavre, @quentinDupont #GRAPOCA

### CODE REVIEW

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                widget.pos.config.meal_voucher_display_product_screen">
+                <i class="fa fa-lg fa-cutlery"/>
+            </t>
+        </t>
+    </t>
+
+    <t t-extend="OrderWidget">
+        <t t-jquery="div.line" t-operation="append">
+            <div class="meal-voucher">
+                (<span>Meal Voucher: </span> <span class="value">0.00 €</span>)
```
Comment (apollet) [inline pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:16]: EURO symbol to remove ? Currency is not always euro ;)

File: pos_meal_voucher/static/src/js/screens.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                           _t(".\n\n Clicking 'Confirm' will validate the payment."),
+                    confirm: function() {
+                        // Note: Due to the bad design of the original function
+                        // the check "Large Amount" will be skipped in that case.
+                        self.validate_order("confirm");
+                    },
+                });
+                return false;
+            }
+            return this._super.apply(this, arguments);;
```
Comment (apollet) [inline pos_meal_voucher/static/src/js/screens.js:154]:
```suggestion
            return this._super.apply(this, arguments);
```

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                    <thead>
+                        <tr>
+                            <th colspan="2">Meal Voucher</th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                        <tr>
+                            <td>Total Eligible</td>
+                            <td>
+                                <span id="meal-voucher-eligible-amount">0.00 €</span>
```
Comment (apollet) [inline pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:46]: EURO symbol to remove ? Currency is not always euro ;)

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                                <i id="meal-voucher-eligible-warning" class="fa fa-warning"/>
+                            </td>
+                        </tr>
+                <t t-if="widget.pos.config.max_meal_voucher_amount !== 0">
+                        <tr>
+                            <td>Max Amount</td>
+                            <td>
+                                <span id="meal-voucher-max-amount">0.00 €</span>
```
Comment (apollet) [inline pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:54]: EURO symbol to remove ? Currency is not always euro ;)

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                                <i id="meal-voucher-max-warning" class="fa fa-warning"/>
+                            </td>
+                        </tr>
+                </t>
+                        <tr>
+                            <td>Total Received</td>
+                            <td>
+                                <span id="meal-voucher-received-amount">0.00 €</span>
```
Comment (apollet) [inline pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:62]: EURO symbol to remove ? Currency is not always euro ;)

File: pos_meal_voucher/static/src/js/screens.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                return;
+            }
+            // Update meal voucher summary
+            var total_eligible = order.get_total_meal_voucher_eligible();
+            this.el.querySelector("#meal-voucher-eligible-amount").textContent = this.format_currency(total_eligible);
+
+            var max_amount = this.pos.config.max_meal_voucher_amount;
+            if (max_amount !== 0) {
+                this.el.querySelector("#meal-voucher-max-amount").textContent = this.format_currency(max_amount);
+            }
```
Comment (apollet) [inline pos_meal_voucher/static/src/js/screens.js:106]: Why only if max_amount != 0 ? You may should let the format_currency even if there is no max_amount.

File: pos_meal_voucher/static/src/js/models.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            res.statement_note = this.statement_note;
+            return res;
+        },
+
+        is_meal_voucher: function() {
+            return (
+                this.manual_meal_voucher === true ||
+                ["paper", "dematerialized"].indexOf(
+                    this.cashregister.journal.meal_voucher_type) !== -1
+                )
```
Comment (apollet) [inline pos_meal_voucher/static/src/js/models.js:72]:
```suggestion
                );
```

**Review by apollet — COMMENTED** [review]
LGTM (not tested), just few changes needed to be perfect to me.
Good job ;)

**Review by remi-filament — APPROVED** [review]
Tested OK on runbot, thanks @legalsylvain for this necessary feature !

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml (hunk 1 above)
Comment (legalsylvain) [reply pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:16]: Just a placeholder. will not be used.

See Odoo code : https://github.com/odoo/odoo/blob/12.0/addons/point_of_sale/static/src/xml/pos.xml#L1321

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml (hunk 4 above)
Comment (legalsylvain) [reply pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:54]: Just a placeholder. will not be used.

See Odoo code : https://github.com/odoo/odoo/blob/12.0/addons/point_of_sale/static/src/xml/pos.xml#L1321

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml (hunk 5 above)
Comment (legalsylvain) [reply pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:62]: Just a placeholder. will not be used.

See Odoo code : https://github.com/odoo/odoo/blob/12.0/addons/point_of_sale/static/src/xml/pos.xml#L1321

File: pos_meal_voucher/static/src/js/screens.js (hunk 6 above)
Comment (legalsylvain) [reply pos_meal_voucher/static/src/js/screens.js:106]: the field is hidden if max_amount === 0, so no need to format the value.

File: pos_meal_voucher/static/src/xml/pos_meal_voucher.xml (hunk 3 above)
Comment (legalsylvain) [reply pos_meal_voucher/static/src/xml/pos_meal_voucher.xml:46]: Just a placeholder. will not be used.

See Odoo code : https://github.com/odoo/odoo/blob/12.0/addons/point_of_sale/static/src/xml/pos.xml#L1321

**Comment by legalsylvain** [comment]
Hi @apollet :  thanks for your review. I answered your question. let me know if something is wrong.

Hi @remi-filament : thanks for your review !

**Review by vvrossem — APPROVED** [review]
LGTM ! nice work :)

**Comment by houssine78** [comment]
Hi @legalsylvain I think that this PR is ready to be merged !

**Comment by legalsylvain** [comment]
/ocabot merge patch

thanks for your reviews.

**Pull request merged**